### PR TITLE
Issue 1143 Implement CAS operations on clustered store

### PR DESCRIPTION
**Pull request by albinsuresh**

### CODE REVIEW

File: clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/ClusteredStore.java [hunk 1]
```diff
@@ -171,20 +194,63 @@ public boolean remove(final K key) throws StoreAccessException {
...
-    // TODO: Make appropriate ServerStoreProxy call
-    throw new UnsupportedOperationException("Implement me");
+    replaceObserver.begin();
+    ReplaceOperation<K, V> operation = new ReplaceOperation<K, V>(key, value);
+    ByteBuffer payload = codec.encode(operation);
+    Chain chain = storeProxy.getAndAppend(key.hashCode(), payload);
+    ResolvedChain<K, V> resolvedChain = resolver.resolve(chain, key);
+    Result<V> result = resolvedChain.getResolvedResult(key);
+    if(result == null) {
+      replaceObserver.end(StoreOperationOutcomes.ReplaceOutcome.REPLACED);
```
Comment (ljacomet) [inline clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/ClusteredStore.java]: Inverted stat outcome

**Comment by lorban** [comment]
LGTM

File: clustered/client/src/test/java/org/ehcache/clustered/client/internal/store/ClusteredStoreTest.java [hunk 2]
```diff
@@ -143,4 +143,42 @@ public void testClear() throws Exception {
...
+    validateStat(store, StoreOperationOutcomes.ConditionalRemoveOutcome.REMOVED, 1);
+    validateStat(store, StoreOperationOutcomes.ConditionalRemoveOutcome.MISS, 2);
+  }
+
+  @Test
+  public void testReplace() throws Exception {
+    assertThat(store.replace(1L, "one"), nullValue());
+    store.put(1L, "one");
+    assertThat(store.replace(1L, "another one").value(), is("one"));
+    validateStat(store, StoreOperationOutcomes.ReplaceOutcome.REPLACED, 1);
```
Comment (ljacomet) [inline clustered/client/src/test/java/org/ehcache/clustered/client/internal/store/ClusteredStoreTest.java]: This _bulk_ validation of stats is why you missed the inverted result in code. Make sure you validate stats immediately after the operation invocation instead.

Applies to all added tests.

**Comment by albinsuresh** [comment]
Addressed the feedback on stats.

**Pull request merged**